pyvista/core/composite.py

```diff
@@ -460,23 +463,40 @@ def __setitem__(self, index: Union[Tuple[int, Optional[str]], int, str], data: D
             i, name = cast(int, index), None
         if data is not None and not is_pyvista_dataset(data):
             data = wrap(data)
+
         if i == -1:
             self.append(data)
             i = self.n_blocks - 1
         else:
+            # this is the only spot in the class where we actually add
+            # data to the MultiBlock
+
+            # check if we are overwriting a block
+            existing_dataset = self.GetBlock(i)
+            if existing_dataset is not None:
+                self._remove_ref(i)
+
             self.SetBlock(i, data)
+            if data is not None:
+                self._refs[data.memory_address] = data
+
         if name is None:
             name = f'Block-{i:02}'
-        self.set_block_name(i, name) # Note that this calls self.Modified()
-        if data not in self.refs:
-            self.refs.append(data)
+        self.set_block_name(i, name)  # Note that this calls self.Modified()
 
     def __delitem__(self, index: Union[int, str]):
         """Remove a block at the specified index."""
         if isinstance(index, str):
             index = self.get_index_by_name(index)
+        self._remove_ref(index)
         self.RemoveBlock(index)
```

Comment:
Also unrelated to this PR, but the current wonky API is also wonky here: `multi[-1] = mesh` appends, but `del multi[-1]` is an error because `RemoveBlock` doesn't accept `-1` as an index. We should make this sane too if/when we redesign the MultiBlock API.